### PR TITLE
src: introduce Nit Static Analysis Framework

### LINKED ISSUE #511: Wanted: A static analysis framework

### PR DESCRIPTION
**Pull request by Morriar**
This is the first step in the introduction of a static analsysis framework for the nit language.

This PR introduces the basic behaviors of a StaticAnalysis and ForwardAnalysis.
I added ReachingDefAnalysis as an example of StaticAnalysis implementation.
I also added `nitsaf`, a client example for the saf module.

I suggest to read this PR commit by commit, I splat it in steps:
* introduce base services
* handle ifs constructs
* handle loops constructs
* handle method declaration

Next steps/PRs will be:
* introduce inter-procedural analysis
* use RTA for better inter-procedural analysis precision
* backward analysis
* add more analysis (isset checks, attr isset, unused properties, useless nullables...)

Closes #511.

### CODE REVIEW

File: lib/core/collection/hash_collection.nit [hunk 1]
```diff
@@ -272,6 +272,12 @@ class HashMap[K, V]
 		enlarge(0)
 	end
 
+	# Build a list filled with the items of `coll`.
+	init from(coll: Map[K, V]) do
+		init
+		for k, v in coll do self[k] = v
+	end
```
Comment (privat) [inline lib/core/collection/hash_collection.nit]: `Map::recover_with` does the loop part.

https://github.com/nitlang/nit/blob/master/lib/core/collection/abstract_collection.nit#L677

File: src/saf/saf_base.nit [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+		var n_block = loop_block
+		if not n_block == null then loop_fix_point(v, n_block)
+	end
+
+	# The block contained by this loop.
+	fun loop_block: nullable ANode is abstract
+end
+
+redef class ADoExpr
+	super ALoopHelper
```
Comment (privat) [inline src/saf/saf_base.nit]: A `do` does not repeat. I dont get this part

**Comment by privat** [comment]
really nice. Some constructions are missing you should list them somewhere, is the doc of the module for instance.

What I think are:

* flow breaks: AReturnExpr, AEscapeExpr (ie ABreakExpr and  AContinueExpr) and AAbortExpr
* other conditionnals: AIfexprExpr, AAssertExpr, ABinBoolExpr (ie AOrExpr, AAndExpr, AOrElseExpr  and AImpliesExpr)

**Comment by privat** [comment]
+1

**Comment by privat** [comment]
please check tests

**Comment by privat** [comment]
ping @Morriar so I can merge

**Comment by jpages** [comment]
Good work :)

+1

File: src/saf/saf_base.nit (hunk 2 above)
Comment (jpages) [inline src/saf/saf_base.nit]: Yes but it has a block inside with a sequence of instructions, I think this is similar to a loop (from a factorization point of view)

**Comment by Morriar** [comment]
Reroll
* added TODOs concerning @jpages remarks (good remarks, will be done in next PRs)
* better documentation regarding @privat concerns (here again, this will come later)

**Comment by privat** [comment]
tests still fail. I'm not sure why

**Comment by Morriar** [comment]
test this please

**Comment by Morriar** [comment]
test this please

**Comment by Morriar** [comment]
Reroll, I added the copyrights in the test files but forgot to update the line numbers in the expected outputs.

**Comment by Morriar** [comment]
Reroll: added nitsaf to skip list of niti and nitvm (timeout).

**Pull request merged**